File: 0198-house-robber/0198-house-robber.py
class Solution:
    def rob(self, nums: List[int]) -> int:
        n = len(nums)
        if n < 3:
            return max(nums)
        
        dp = [0]*n
        
        dp[0] = nums[0]
        dp[1] = nums[1]
        dp[2] = nums[0] + nums[2]
        
        for i in range(3, n):
            dp[i] = max(dp[i-1], dp[i-2]+nums[i], dp[i-3]+nums[i])
        
        return max(dp[n-1], dp[n-2])

File: 0198-house-robber/0198-house-robber.rs
impl Solution {
    pub fn rob(nums: Vec<i32>) -> i32 {
        let n : usize = nums.len();
        let mut dp = vec![0; n + 1];
        dp[0] = nums[0];
        if n == 1 {
            return dp[0];
        }
        dp[1] = nums[1].max(nums[0]);
        for i in 2..n {
            dp[i] = dp[i-1].max(dp[i-2] + nums[i]);
        }
        dp[n-1].max(dp[n-2])
    }
}